beacon-chain/sync/data_column_sidecars.go

```diff
@@ -39,249 +39,394 @@ type DataColumnSidecarsParams struct {
-		block := roBlock.Block()
-		commitments, err := block.Body().BlobKzgCommitments()
-		if err != nil {
-			return nil, errors.Wrapf(err, "get blob kzg commitments for block root %#x", roBlock.Root())
-		}
-		if len(commitments) == 0 {
-			continue
-		}
+	initialMissingRootCount := len(incompleteRoots)
 
-		slotsWithCommitments[block.Slot()] = true
-		root := roBlock.Root()
+	// Request sidecars from storage (by reconstructing them from other available sidecars if needed).
+	result, err := requestSidecarsFromStorage(params.Storage, incompleteRoots, storedIndicesByRoot, requestedIndices)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "request sidecars from storage")
+	}
 
-		// Step 1: Get the requested sidecars for this root if available in storage
-		requestedColumns, err := tryGetStoredColumns(params.Storage, root, indices)
-		if err != nil {
-			return nil, errors.Wrapf(err, "try get direct columns for root %#x", root)
-		}
-		if requestedColumns != nil {
-			result[root] = requestedColumns
-			continue
-		}
+	log := log.WithField("initialMissingRootCount", initialMissingRootCount)
 
-		// Step 2: If step 1 failed, reconstruct the requested sidecars from what is available in storage
-		requestedColumns, err = tryGetReconstructedColumns(params.Storage, root, indices)
-		if err != nil {
-			return nil, errors.Wrapf(err, "try get reconstructed columns for root %#x", root)
-		}
-		if requestedColumns != nil {
-			result[root] = requestedColumns
-			continue
-		}
+	if len(incompleteRoots) == 0 {
+		log.WithField("finalMissingRootCount", len(incompleteRoots)).Debug("Fetched data column sidecars from storage")
+		return result, nil, nil
+	}
+
+	// Request direct sidecars from peers.
+	directSidecarsByRoot, err := requestDirectSidecarsFromPeers(params, roBlocks, slotsWithCommitments, storedIndicesByRoot, requestedIndices, incompleteRoots)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "request direct sidecars from peers")
+	}
 
-		// Step 3a: If steps 1 and 2 failed, keep track of the sidecars that need to be queried from peers
-		//          and those that are already stored.
-		indicesToQueryMap, indicesStoredMap := categorizeIndices(params.Storage, root, indices)
+	// Merge sidecars in storage and those received from peers. Reconstruct if needed.
+	mergedSidecarsByRoot, err := mergeStorageAndInputs(params.Storage, storedIndicesByRoot, directSidecarsByRoot, requestedIndices, incompleteRoots)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "try merge storage and mandatory inputs")
+	}
 
-		if len(indicesToQueryMap) > 0 {
-			missingIndicesByRoot[root] = indicesToQueryMap
-		}
-		if len(indicesStoredMap) > 0 {
-			indicesByRootStored[root] = indicesStoredMap
-		}
+	for root, sidecars := range mergedSidecarsByRoot {
+		result[root] = sidecars
 	}
 
-	// Early return if no sidecars need to be queried from peers.
-	if len(missingIndicesByRoot) == 0 {
-		return result, nil
+	if len(incompleteRoots) == 0 {
```

Comment:
this should go above line 104 through I think

Reply:
https://github.com/OffchainLabs/prysm/blob/4f5661aca2437f2d09246d2e148d9c9a398da46f/beacon-chain/sync/data_column_sidecars.go#L104-L111

We need sidecars in `mergedSidecarsByRoot` to be injected in `result` **before** exiting the function (even if `incompleteRoots` is empty).